routing/routing.go

```diff
@@ -140,10 +141,76 @@ type RouteFilter struct {
 	Index int
 }
 
+// LBMetrics contains metrics used by LB algorithms
+type LBMetrics struct {
+	inflightRequests int32
+}
+
+// IncInflightRequest increments the number of outstanding requests from the proxy to a given backend.
+func (m *LBMetrics) IncInflightRequest() {
+	atomic.AddInt32(&m.inflightRequests, 1)
+}
+
+// DecInflightRequest decrements the number of outstanding requests from the proxy to a given backend.
+func (m *LBMetrics) DecInflightRequest() {
+	atomic.AddInt32(&m.inflightRequests, -1)
+}
+
+// GetInflightRequests decrements the number of outstanding requests from the proxy to a given backend.
+func (m *LBMetrics) GetInflightRequests() int {
+	return int(atomic.LoadInt32(&m.inflightRequests))
+}
+
 // LBEndpoint represents the scheme and the host of load balanced
 // backends.
 type LBEndpoint struct {
 	Scheme, Host string
+	Metrics      *LBMetrics
+}
+
+type lbEndpointCollection struct {
+	endpoints    []LBEndpoint
+	endpointsMap map[string]LBEndpoint
+}
+
+// lbEndpointCollection represents a collection of endpoints.
+// The underlying endpoints can be accessed using public getters and setters.
+// Internally endpoints are stored in two synchronized data structures, so
+// it is possible to reference them by index or key.
+func NewEndpointCollection(r *Route) (*lbEndpointCollection, error) {
+	collection := &lbEndpointCollection{
+		endpoints:    make([]LBEndpoint, 0, len(r.Route.LBEndpoints)),
+		endpointsMap: map[string]LBEndpoint{},
+	}
+	for _, e := range r.Route.LBEndpoints {
+		eu, err := url.ParseRequestURI(e)
+		if err != nil {
+			return nil, err
+		}
+		collection.Set(eu.Host, LBEndpoint{Scheme: eu.Scheme, Host: eu.Host, Metrics: &LBMetrics{}})
+	}
+	return collection, nil
+}
+
+// Stores an endpoint indexed by key
+func (e *lbEndpointCollection) Set(key string, item LBEndpoint) {
+	e.endpoints = append(e.endpoints, item)
```

Comment:
this potentially can lead to inconsistent internal state, when the key already exists in the map. If they key exists, we should find the currently stored item's index in the slice, and replace the item in the slice instead of appending.

Another approach, and maybe better, would be if we didn't export the Set method, and keep the collection immutable, only populated by the NewEndpointCollection initializer.

WDYT?